### PR TITLE
drivers:adc: Add ad405x no-OS drivers

### PR DESCRIPTION
**Pull request by ribdp**
## Pull Request Description

1. Add support for sample, averaging and burst averaging modes
2. Add support for spi read-write
3. Add bitfield definitions and apis for modifying and accessing bitfields

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [ ] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

### CODE REVIEW

File: drivers/adc/ad405x/ad405x.c [hunk 1]
```diff
@@ -0,0 +1,900 @@
...
+/**
+ * @brief Toggle the CNV pin to start a conversion.
+ * @param dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad405x_convst(struct ad405x_dev *dev)
+{
+	int ret;
+
+	ret = no_os_gpio_set_value(dev->gpio_cnv, 1);
```
Comment (amiclaus) [inline drivers/adc/ad405x/ad405x.c:350]: use please https://github.com/analogdevicesinc/no-OS/blob/3cb5f58f6815c4fa63b732bb5072d42483ad6f81/include/no_os_gpio.h#L119

File: drivers/adc/ad405x/ad405x.c [hunk 2]
```diff
@@ -0,0 +1,900 @@
...
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad405x_read_val(struct ad405x_dev *dev, int32_t *data)
+{
+	int ret;
+	int32_t data_read;
+
+	ret = ad405x_convst(dev);
+	if (ret)
+		return -1;
```
Comment (amiclaus) [inline drivers/adc/ad405x/ad405x.c:423]: `einval`

File: drivers/adc/ad405x/ad405x.c [hunk 3]
```diff
@@ -0,0 +1,900 @@
...
+
+	ret = ad405x_spi_data_read(dev, &data_read);
+	if (!ret)
```
Comment (amiclaus) [inline drivers/adc/ad405x/ad405x.c:426]: what happens if ret != 0 ? What will the function return?

File: drivers/adc/ad405x/ad405x.c [hunk 4]
```diff
@@ -0,0 +1,900 @@
...
+
+	return 0;
+}
+
+/**
+ * @brief Get sample rate for Burst and Autonomous Modes
+ * @param dev - The device structure.
+ * @return Sample rate value
+ */
+enum ad405x_sample_rate ad405x_get_sample_rate(struct ad405x_dev *dev)
```
Comment (amiclaus) [inline drivers/adc/ad405x/ad405x.c:463]: is this function needed?

File: drivers/adc/ad405x/ad405x.c [hunk 5]
```diff
@@ -0,0 +1,900 @@
...
+	int ret;
+
+	if (!dev)
+		return - EINVAL;
+
+	/* Restrict filter length depending on active device selected */
+	switch (dev->active_device) {
+	case ID_AD4050:
+		if (length > AD405X_LENGTH_256)
+			return -EINVAL;
```
Comment (amiclaus) [inline drivers/adc/ad405x/ad405x.c:489]: where is the `break` statement here?

File: drivers/adc/ad405x/ad405x.c [hunk 6]
```diff
@@ -0,0 +1,900 @@
...
+
+/**
+ * @brief Get DEV_EN signal polarity
+ * @param dev - The device structure.
+ * @return DEV_EN signal polarity
+ */
+enum ad405x_dev_en_polarity ad405x_get_dev_en_polarity(struct ad405x_dev *dev)
+{
+	if (!dev)
+		return - EINVAL;
```
Comment (amiclaus) [inline drivers/adc/ad405x/ad405x.c:556]: drop space between `-` and `EINVAL`. Check throughout the entire code.

File: drivers/adc/ad405x/ad405x.c [hunk 7]
```diff
@@ -0,0 +1,900 @@
...
+	return 0;
+}
+
+/**
+ * @brief Get output data format setting.
+ * @param dev - The device structure.
+ * @return DATA_FORMAT setting or negative error code.
+ */
+enum ad405x_out_data_format ad405x_get_data_format(
+	struct ad405x_dev *dev)
```
Comment (amiclaus) [inline drivers/adc/ad405x/ad405x.c:731]: same here: is this function really needed?

File: drivers/adc/ad405x/ad405x.c [hunk 8]
```diff
@@ -0,0 +1,900 @@
...
+
+
+	ret = no_os_spi_remove(dev->spi_desc);
+	if (ret)
+		return ret;
+
+	free(dev);
+
+	return ret;
+}
```
Comment (amiclaus) [inline drivers/adc/ad405x/ad405x.c:900]: new line missing.

File: drivers/adc/ad405x/ad405x.h [hunk 9]
```diff
@@ -0,0 +1,400 @@
+/*******************************************************************************
+ *   @file   ad405x.h
+ *   @brief  Header file of AD405X Driver.
+ *   @author George Mois (george.mois@analog.com)
```
[the remark this answers is not in the document]
Comment (ribdp) [reply drivers/adc/ad405x/ad405x.h:4]: The initial driver was done by George. I generalized it for other generics besides AD4052, and added support for the averaging mode. George suggested I add my name also, to the authors list

File: drivers/adc/ad405x/ad405x.c [hunk 10]
```diff
@@ -0,0 +1,900 @@
...
+	if (dev->operation_mode != mode) {
+		if (dev->operation_mode != AD405X_CONFIG_MODE_OP) {
+			ret = ad405x_exit_command(dev);
+			if (ret)
+				return ret;
+		}
+
+		switch (mode) {
+		case AD405X_CONFIG_MODE_OP:
+			return 0;
```
[the remark this answers is not in the document]
Comment (ribdp) [reply drivers/adc/ad405x/ad405x.c:328]: This is done in the `ad405x_exit_command` function which is called on Line 320. 
If the user tries to set the same operating mode, nothing needs to be done. 
If the user tries to switch to a different mode, they'd first need to send the exit command to get out of the first mode back into the config mode.

File: drivers/adc/ad405x/ad405x.c (hunk 4 above)
Comment (ribdp) [reply drivers/adc/ad405x/ad405x.c:463]: Not needed. Will remove

File: drivers/adc/ad405x/ad405x.c (hunk 7 above)
Comment (ribdp) [reply drivers/adc/ad405x/ad405x.c:731]: Not needed. Will remove

**Comment by ribdp** [comment]
Changelog - 
1. Update copyright year in both files
2. Add missing new line at the end of files
3. Fix typo in function description in `ad405x_set_config_mode`
4. Remove unused functions 
5. Add missing break statement in `ad405x_set_averaging_filter_length`
6. Fix return logic in `ad405x_read_val`
7. Add static const definition for reset patter bytes buffer

**Comment by ribdp** [comment]
Changelog - 
1. Rebased with main
2. Added readme documentation commit

**Comment by ribdp** [comment]
> please add also a commit for the documentation, here is an example [f3c60c0](https://github.com/analogdevicesinc/no-OS/commit/f3c60c051f46fabea38ebe2a06a90900e3997357)

Added the documentation now

**Comment by ribdp** [comment]
Do the PR changes look okay now?

**Review by amiclaus — APPROVED** [review]

**Review by danmois — APPROVED** [review]

**Pull request merged**